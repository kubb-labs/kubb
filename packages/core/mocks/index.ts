import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
