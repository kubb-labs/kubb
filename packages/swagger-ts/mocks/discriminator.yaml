info:
  title: Swagger Petstore
openapi: 3.1.0
components:
  schemas:
    Cat:
      title: Cat
      type: object
      properties:
        id:
          format: uuid
          title: Id
          type: string
        type:
          const: Cat
          title: Type
        meow:
          type: boolean
          title: Meowing prevalence
      required:
      - id
      - type
    Dog:
      title: Dog
      type: object
      properties:
        id:
          format: uuid
          title: Id
          type: string
        bark:
          type: number
          title: Barking frequency
        type:
          const: Dog
          title: Type
      required:
      - id
      - type
    Petstore:
      title: Petstore
      type: object
      description: A project is an analysis project. It has a name, an id, and has
        a dataset
      properties:
        id:
          title: Id
          type: integer
        pets:
          default: []
          items:
            discriminator:
              mapping:
                Cat: "#/components/schemas/Cat"
                Dog: "#/components/schemas/Dog"
              propertyName: type
            oneOf:
            - "$ref": "#/components/schemas/Cat"
            - "$ref": "#/components/schemas/Dog"
          title: Pets
          type: array
      required:
      - id

paths:
