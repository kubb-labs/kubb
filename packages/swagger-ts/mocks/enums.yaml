
definitions:
  enumVarNames.Type:
    enum:
    - 0
    - 1
    type: integer
    x-enum-varnames:
    - Pending
    - Received
  enumNames.Type:
    enum:
    - 0
    - 1
    x-enumNames:
      - Pending
      - Received
  enum.String:
    type: string
    enum:
      - created_at
      - description
      - FILE.UPLOADED
      - FILE.DOWNLOADED
  enum.InObject:
    type: object
    properties:
      reason:
        type: string
        enum:
          - created_at
          - description
  enum.AllOf:
    type: object
    properties:
      reason:
        allOf:
          - type: string
            enum:
              - created_at
              - description

  enum.Array:
    type: object
    properties:
     identifier:
        type: array
        prefixItems:
          - type: integer
            description: Number of items
          - type: string
            description: Description
          - enum:
              - "NW"
              - "NE"
              - "SW"
              - "SE"
            description: Wind direction

  enum.Items:
    type: array
    minItems: 1
    uniqueItems: true
    items:
      type: string
      enum:
        - created_at
        - description
swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumVarNames.Type'
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNames.Type'
