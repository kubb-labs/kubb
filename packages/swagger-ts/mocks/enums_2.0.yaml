definitions:
  enum.Array:
    type: object
    properties:
      identifier:
        type: array
        prefixItems:
          - type: integer
            description: Number of items
          - type: string
            description: Description
          - enum:
              - "NW"
              - "NE"
              - "SW"
              - "SE"
            description: Wind direction

swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths: {}
