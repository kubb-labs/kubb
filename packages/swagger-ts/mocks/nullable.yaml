
definitions:
  Nullable:
    type: object
    properties:
      foo:
        type: string
        nullable: true
openapi: 3.1.0
info:
  title: test
  version: "1.0.0"
paths:
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/Nullable'
