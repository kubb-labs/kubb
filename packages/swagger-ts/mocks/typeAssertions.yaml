info:
  title: Type Assertions
openapi: 3.1.0
paths: {}
components:
  schemas:
    Body_upload_file_api_assets_post:
      properties:
        file:
          format: binary
          title: File
          type: string
      required:
      - file
      title: Body_upload_file_api_assets_post
      type: object
      $comment: |
        export type BodyUploadFileApiAssetsPost = {
            /**
            * @type string binary
            */
            file: Blob;
        };
    Plain_file:
      format: binary
      title: File
      type: string
      $comment: export type PlainFile = Blob;
    Plain_date:
      format: date
      title: Date
      type: string
      $comment: export type PlainDate = Date;
    Plain_time:
      format: time
      title: Time
      type: number
      $comment: export type PlainTime = number;
    Plain_date_time:
      format: date-time
      title: Datetime
      type: string
      $comment:
    Plain_email:
      format: email
      title: Email
      type: string
    Plain_uuid:
      format: uuid
      title: UUID
      type: string

