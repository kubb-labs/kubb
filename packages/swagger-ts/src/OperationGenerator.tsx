import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OasType } from './components/OasType.tsx'
import { OperationSchema } from './components/OperationSchema.tsx'

import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[]): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        {plugin.options.oasType && <OasType.File name="oas" typeName="Oas" />}
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  async operation(operation: Operation, options: PluginOptions['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })
    root.render(
      <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Oas.Operation operation={operation}>
          <OperationSchema.File mode={mode} />
        </Oas.Operation>
      </Oas>,
      { meta: { pluginManager, plugin: { ...plugin, options } } },
    )

    return root.files
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }
  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
