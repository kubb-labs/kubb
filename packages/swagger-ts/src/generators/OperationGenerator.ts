import { getRelativePath } from '@kubb/core/utils'
import { print } from '@kubb/parser'
import * as factory from '@kubb/parser/factory'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'
import { camelCase, pascalCase } from 'change-case'

import { TypeBuilder } from '../builders/index.ts'

import type { KubbFile } from '@kubb/core'
import type { FileResolver, Operation, OperationSchemas, Resolver } from '@kubb/swagger'
import type ts from 'typescript'
import type { FileMeta, Options as PluginOptions } from '../types.ts'

type Options = {
  usedEnumNames: Record<string, number>

  mode: KubbFile.Mode
  enumType: NonNullable<PluginOptions['enumType']>
  dateType: NonNullable<PluginOptions['dateType']>
  optionalType: NonNullable<PluginOptions['optionalType']>
}

export class OperationGenerator extends Generator<Options> {
  resolve(operation: Operation): Resolver {
    const { pluginManager, plugin } = this.context

    return resolve({
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin?.key,
    })
  }

  async all(): Promise<KubbFile.File | null> {
    return null
  }

  #printCombinedSchema(name: string, operation: Operation, schemas: OperationSchemas): string {
    const properties: Record<string, ts.TypeNode> = {
      'response': factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.response.name),
        undefined,
      ),
    }

    if (schemas.request) {
      properties['request'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.request.name),
        undefined,
      )
    }

    if (schemas.pathParams) {
      properties['pathParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.pathParams.name),
        undefined,
      )
    }

    if (schemas.queryParams) {
      properties['queryParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.queryParams.name),
        undefined,
      )
    }

    if (schemas.headerParams) {
      properties['headerParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.headerParams.name),
        undefined,
      )
    }

    if (schemas.errors) {
      properties['errors'] = factory.createUnionDeclaration({
        nodes: schemas.errors.map(error => {
          return factory.createTypeReferenceNode(
            factory.createIdentifier(error.name),
            undefined,
          )
        }),
      })!
    }

    const namespaceNode = factory.createNamespaceDeclaration({
      name: operation.method === 'get' ? `${name}Query` : `${name}Mutation`,
      statements: Object.keys(properties).map(key => {
        const type = properties[key]
        if (!type) {
          return undefined
        }
        return factory.createTypeAliasDeclaration({
          modifiers: [factory.modifiers.export],
          name: pascalCase(key),
          type,
        })
      }).filter(Boolean),
    })

    return print(namespaceNode)
  }

  async get(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { mode, enumType, dateType, optionalType, usedEnumNames } = options
    const { pluginManager, plugin } = this.context

    const type = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: type.baseName, pluginKey: plugin?.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin?.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new TypeBuilder({
      usedEnumNames,
      fileResolver: mode === 'file' ? undefined : fileResolver,
      withJSDocs: true,
      resolveName: (params) => pluginManager.resolveName({ ...params, pluginKey: plugin?.key }),
      enumType,
      optionalType,
      dateType,
    })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    const combinedSchemaSource = this.#printCombinedSchema(type.name, operation, schemas)

    return {
      path: type.path,
      baseName: type.baseName,
      source: [source, combinedSchemaSource].join('\n'),
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async post(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { mode, enumType, dateType, optionalType, usedEnumNames } = options
    const { pluginManager, plugin } = this.context

    const type = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: type.baseName, pluginKey: plugin?.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin?.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new TypeBuilder({
      usedEnumNames,
      fileResolver: mode === 'file' ? undefined : fileResolver,
      withJSDocs: true,
      resolveName: (params) => pluginManager.resolveName({ ...params, pluginKey: plugin?.key }),
      enumType,
      optionalType,
      dateType,
    })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.request)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    const combinedSchemaSource = this.#printCombinedSchema(type.name, operation, schemas)

    return {
      path: type.path,
      baseName: type.baseName,
      source: [source, combinedSchemaSource].join('\n'),
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async put(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async patch(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async delete(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
}
