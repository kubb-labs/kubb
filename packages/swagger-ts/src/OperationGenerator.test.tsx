import { FileManager } from '@kubb/core'
import { mockedPluginManager } from '@kubb/core/mocks'
import { OasManager } from '@kubb/swagger'

import { OperationGenerator } from './OperationGenerator.tsx'

import type { KubbFile } from '@kubb/core'
import type { Plugin } from '@kubb/core'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from './types.ts'

describe('OperationGenerator', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-ts/mocks/petStore.yaml' },
  })
  test('[GET] /pets with optionalType `questionToken`', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
      oasType: false,
      unknownType: 'any',
    }

    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      override: undefined,
    })
    const operation = oas.operation('/pets', 'get')
    const operationShowById = oas.operation('/pets/{petId}', 'get')

    const files = (await og.operation(operation, options)) as KubbFile.File[]
    const getShowByIdFiles = (await og.operation(operationShowById, options)) as KubbFile.File[]

    files.forEach((file) => {
      expect(FileManager.getSource(file)).toMatchSnapshot()
    })

    getShowByIdFiles.forEach((file) => {
      expect(FileManager.getSource(file)).toMatchSnapshot()
    })
  })

  test('[POST] /pets 201', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
      oasType: false,
      unknownType: 'any',
    }

    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      override: undefined,
    })
    const operation = oas.operation('/pets', 'post')
    const files = (await og.operation(operation, options)) as KubbFile.File[]

    files.forEach((file) => {
      expect(FileManager.getSource(file)).toMatchSnapshot()
    })
  })

  test('[DELETE] /pet/{petId} with unknownType `any`', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
      oasType: false,
      unknownType: 'any',
    }

    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      override: undefined,
    })
    const operation = oas.operation('/pet/{petId}', 'delete')
    const files = (await og.operation(operation, options)) as KubbFile.File[]

    files.forEach((file) => {
      expect(FileManager.getSource(file)).toMatchSnapshot()
    })
  })
  test('[DELETE] /pet/{petId} with unknownType `unknown`', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
      oasType: false,
      unknownType: 'unknown',
    }

    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      override: undefined,
    })
    const operation = oas.operation('/pet/{petId}', 'delete')
    const files = (await og.operation(operation, options)) as KubbFile.File[]

    files.forEach((file) => {
      expect(FileManager.getSource(file)).toMatchSnapshot()
    })
  })
})
