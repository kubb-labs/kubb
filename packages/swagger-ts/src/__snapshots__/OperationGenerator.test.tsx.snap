// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OperationGenerator > [DELETE] /pet/{petId} with unknownType \`any\` 1`] = `
"export type DeletePetPetidMutationResponse = any;

 export type DeletePetPetidMutation = {
    Response: DeletePetPetidMutationResponse;
};
"
`;

exports[`OperationGenerator > [DELETE] /pet/{petId} with unknownType \`unknown\` 1`] = `
"export type DeletePetPetidMutationResponse = unknown;

 export type DeletePetPetidMutation = {
    Response: DeletePetPetidMutationResponse;
};
"
`;

exports[`OperationGenerator > [GET] /pets with optionalType \`questionToken\` 1`] = `
"import { Error } from "./Error";
import type { Pets } from "./Pets";

 export type ListPetsQueryParams = {
    /**
     * @description How many items to return at one time (max 100)
     * @type string | undefined
    */
    limit?: string;
};

 /**
 * @description A paged array of pets
*/
export type ListPets200 = Pets;

 /**
 * @description unexpected error
*/
export type ListPetsError = Error;

 /**
 * @description A paged array of pets
*/
export type ListPetsQueryResponse = Pets;

 export type ListPetsQuery = {
    Response: ListPetsQueryResponse;
    QueryParams: ListPetsQueryParams;
};
"
`;

exports[`OperationGenerator > [GET] /pets with optionalType \`questionToken\` 2`] = `
"import { Error } from "./Error";
import type { Pet } from "./Pet";

 export type ShowPetByIdPathParams = {
    /**
     * @description The id of the pet to retrieve
     * @type string
    */
    petId: string;
    /**
     * @description The id of the pet to retrieve
     * @type string
    */
    testId: string;
};

 /**
 * @description Expected response to a valid request
*/
export type ShowPetById200 = Pet;

 /**
 * @description unexpected error
*/
export type ShowPetByIdError = Error;

 /**
 * @description Expected response to a valid request
*/
export type ShowPetByIdQueryResponse = Pet;

 export type ShowPetByIdQuery = {
    Response: ShowPetByIdQueryResponse;
    PathParams: ShowPetByIdPathParams;
};
"
`;

exports[`OperationGenerator > [POST] /pets 201 1`] = `
"import type { Error } from "./Error";
import type { ErrorCode } from "./ErrorCode";

 /**
 * @description Null response
*/
export type CreatePets201 = (Error & {
    name?: ErrorCode;
});

 /**
 * @description unexpected error
*/
export type CreatePetsError = Error;

 export type CreatePetsMutationRequest = {
    /**
     * @type string
    */
    name: string;
    /**
     * @type string
    */
    tag: string;
};

 /**
 * @description Null response
*/
export type CreatePetsMutationResponse = (Error & {
    name?: ErrorCode;
});

 export type CreatePetsMutation = {
    Response: CreatePetsMutationResponse;
    Request: CreatePetsMutationRequest;
};
"
`;
